packages/typescript-estree/src/create-program/useProvidedProgram.ts

```diff
@@ -0,0 +1,87 @@
+import debug from 'debug';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+import { Extra } from '../parser-options';
+import {
+  ASTAndProgram,
+  CORE_COMPILER_OPTIONS,
+  getAstFromProgram,
+} from './shared';
+
+const log = debug('typescript-eslint:typescript-estree:useProvidedProgram');
+
+function useProvidedProgram(
+  programInstance: ts.Program,
+  extra: Extra,
+): ASTAndProgram | undefined {
+  log('Retrieving ast for %s from provided program instance', extra.filePath);
+
+  programInstance.getTypeChecker(); // ensure parent pointers are set in source files
+
+  const astAndProgram = getAstFromProgram(programInstance, extra);
+
+  if (!astAndProgram) {
+    const relativeFilePath = path.relative(
+      extra.tsconfigRootDir || process.cwd(),
+      extra.filePath,
+    );
+    const errorLines = [
+      '"parserOptions.program" has been provided for @typescript-eslint/parser.',
+      `The file was not found in the provided program instance: ${relativeFilePath}`,
+    ];
+
+    throw new Error(errorLines.join('\n'));
+  }
+
+  return astAndProgram;
+}
+
+/**
+ * Utility offered by parser to help consumers construct their own program instance.
+ *
+ * @param configFile the path to the tsconfig.json file, relative to `projectDirectory`
+ * @param projectDirectory the project directory to use as the CWD, defaults to `process.cwd()`
+ */
+function createProgramFromConfigFile(
+  configFile: string,
+  projectDirectory?: string,
+): ts.Program {
+  if (ts.sys === undefined) {
+    throw new Error(
+      '`createProgramFromConfigFile` is only supported in a Node-like environment.',
+    );
+  }
+
+  const parsed = ts.getParsedCommandLineOfConfigFile(
+    configFile,
+    CORE_COMPILER_OPTIONS,
+    {
+      onUnRecoverableConfigFileDiagnostic: diag => {
+        throw new Error(formatDiagnostics([diag])); // ensures that `parsed` is defined.
+      },
+      fileExists: fs.existsSync,
+      getCurrentDirectory: () =>
+        (projectDirectory && path.resolve(projectDirectory)) || process.cwd(),
+      readDirectory: ts.sys.readDirectory,
+      readFile: file => fs.readFileSync(file, 'utf-8'),
+      useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
```

Comment:
if someone wants to use `program` from a browser environment - this will crash as `ts.sys` will be undefined.

it's probs best to just use the variable defined in the shared file

Reply:
See earlier comment forcing Node-like environment - since it's not core functionality, I feel like that's an appropriate requirement. I'm much more concerned about `ts.sys.readDirectory` than I am `ts.sys.useCaseSensitiveFileNames`.

Reply:
Rephrased - you can use `parserOptions.program` from a browser! You'll just have to figure out how to build that program instance yourself.

Reply:
oh sorry... I wasn't thinking in terms of the github UI with how resolved threads get collapsed.
I was thinking in terms of phabricator where the comments are threaded on the file but time-ordered in the activity log.

yeah it's late 😅